mt-kahypar/partition/refinement/fm/global_rollback.cpp

```diff
@@ -235,115 +204,94 @@ namespace mt_kahypar {
+      // distribute gains to pins
+      const HyperedgeWeight we = phg.edgeWeight(e);
+      for (HypernodeID v : phg.pins(e)) {
+        if (tracker.wasNodeMovedInThisRound(v)) {
+          const MoveID m_id = tracker.moveOfNode[v];
+          Move& m = tracker.getMove(m_id);
 
-        if (phg.edgeSize(e_global) > 2) {
-          const HyperedgeID e = phg.nonGraphEdgeID(e_global);
+          const bool benefit = r[m.from].last_out == m_id && r[m.from].first_in > m_id && r[m.from].remaining_pins == 0;
+          const bool penalty = r[m.to].first_in == m_id && r[m.to].last_out < m_id && r[m.to].remaining_pins == 0;
 
-          const MoveID firstInFrom = firstMoveIn(e, m.from);
-          if (remainingPinsFromBeginningOfMovePhase(e, m.from) == 0 && lastMoveOut(e, m.from) == globalMoveID
-              && (firstInFrom > globalMoveID || firstInFrom < firstMoveID)) {
-            gain += edgeWeight;
+          if (benefit && !penalty) {    // only apply update if they're mutually exclusive
+            // increase gain of v by w(e)
+            __atomic_fetch_add(&m.gain, we, __ATOMIC_RELAXED);
           }
 
-          const MoveID lastOutTo = lastMoveOut(e, m.to);
-          if (remainingPinsFromBeginningOfMovePhase(e, m.to) == 0
-              && firstMoveIn(e, m.to) == globalMoveID && lastOutTo < globalMoveID) {
-            gain -= edgeWeight;
+          if (!benefit && penalty) {
+            // decrease gain of v by w(e)
+            __atomic_fetch_sub(&m.gain, we, __ATOMIC_RELAXED);
           }
+        }
+      }
 
-        } else {
-
-          const HypernodeID u = m.node;
-          const HypernodeID v = phg.graphEdgeHead(e_global, u);
-          const MoveID move_id_of_v = tracker.moveOfNode[v];
-          const bool v_moved = !tracker.isMoveStale(move_id_of_v) && tracker.isMoveStillValid(move_id_of_v);
-
-          PartitionID pv;
-          if (v_moved) {
-            const MoveID local_move_id_of_v = move_id_of_v - firstMoveID;
-            const Move& move_of_v = move_order[local_move_id_of_v];
-            if (local_move_id_of_v < localMoveID) {
-              pv = move_of_v.to;
-            } else {
-              pv = move_of_v.from;
-            }
+      if (num_parts <= static_cast<int>(2 * phg.edgeSize(e))) {
+        for (PartitionID i = 0; i < num_parts; ++i) {
+          r[i] = RecalculationData();
+        }
+      } else {
+        for (HypernodeID v : phg.pins(e)) {
+          if (tracker.wasNodeMovedInThisRound(v)) {
+            const Move& m = tracker.getMove(tracker.moveOfNode[v]);
+            r[m.from] = RecalculationData();
+            r[m.to] = RecalculationData();
           } else {
-            pv = phg.partID(v);
+            r[phg.partID(v)].remaining_pins = 0;
           }
+        }
+      }
+    };
 
-          if (pv == m.to) {
-            gain += edgeWeight;
-          } else if (pv == m.from) {
-            gain -= edgeWeight;
+    if (!context.refinement.fm.rollback_sensitive_to_num_moves) {
```

Comment:
weird naming: rollback_sensitive_to_num_moves. Maybe some explanation in context.h would be helpful

Reply:
yup. any alternative suggestions? I thought about calling it iter-moves but that doesn't really explain what it does either